{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Merging BCR & TCR *CITE* and *nonCITE* samples \n",
    "## Input\n",
    "### BCR & TCR *CITE* samples\n",
    "* Folder path: `./covid_cite_data`\n",
    "* Each **cite** has a folder; naming pattern: CITEx_BCR/TCR_VHT_cellranger\n",
    "* **CAUTION**: sample *TP8B* has no corresponding *CITE* sample\n",
    "* Each **cite** has a `filtered_contig_annotations.csv` file; important columns:\n",
    "  * **barcode**: each cell has a unique barcode\n",
    "  * **contig_id**: each contig has a unique contig_id; multiple contigs may belong to one cell\n",
    "* Each sample also has a `filtered_contig.fasta.fasta` file; search sequence by contig_id\n",
    "\n",
    "### Demultiplexed BCR & TCR *CITE* samples\n",
    "* Folder path: `./CITESEQ_demultiplex`\n",
    "* Each sample **each cite** has a folder; naming pattern: sample name_BCR/TCR_CITEx\n",
    "* **CAUTION**: sample *TP8B* has no corresponding *CITE* sample\n",
    "* Each sample has a `filtered_contig_annotations.csv` file; important columns:\n",
    "  * **barcode**: each cell has a unique barcode\n",
    "  * **contig_id**: each contig has a unique contig_id; multiple contigs may belong to one cell\n",
    "* Each sample **NOT** has a `filtered_contig.fasta.fasta` file; so need to search sequence by contig_id in fasta files in the *CITE* folder\n",
    "    \n",
    "### BCR & TCR *nonCITE* samples\n",
    "* Folder path: `./covid_noncite_data`\n",
    "* Each sample has a folder; naming pattern: sample name_BCR/TCR_VHT_cellranger\n",
    "* Each sample has a `filtered_contig_annotations.csv` file; important columns:\n",
    "  * **barcode**: each cell has a unique barcode\n",
    "  * **contig_id**: each contig has a unique contig_id; multiple contigs may belong to one cell\n",
    "* Each sample also has a `filtered_contig.fasta.fasta` file; search sequence by contig_id\n",
    "  \n",
    "## Output\n",
    "* Output folder: `./covid_merged_data`\n",
    "* CAUTION: **ALL** cell **barcodes** together with their **contig ids** in both *csv* and *fasta* are renamed. The new name has a suffix `_nonCITE` or `_CITEx`\n",
    "* Include 2 files:\n",
    "  * `filtered_contig_annotations.csv`\n",
    "  * `filtered_contig.fasta`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import display\n",
    "import pandas as pd\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "import numpy as np\n",
    "from shutil import rmtree"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Specify folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CITE samples folder\n",
    "cite_folder = r'./covid_cite_data'\n",
    "# Demultiplex floder\n",
    "demultiplex_folder = r'./CITESEQ_demultiplex'\n",
    "# nonCITE samples folder\n",
    "noncite_folder = r'./covid_noncite_data'\n",
    "# output folder\n",
    "output_folder = r'./covid_merged_data'\n",
    "if os.path.isdir(output_folder):\n",
    "    rmtree(output_folder)\n",
    "os.mkdir(output_folder)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define functions to meger data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def saveFile(sample, rename_barcodes, output_name):\n",
    "    '''save combined csv and fastq files\n",
    "    sample is a nested dict with 'nonCITE', 'CITEx' as keys\n",
    "    the item of sample is still a dict with 'file path' and 'seq path' as keys\n",
    "    '''\n",
    "    \n",
    "    to_write_csv = []\n",
    "    to_write_fastq = []\n",
    "    \n",
    "    for k, v in sample.items():\n",
    "        tmp_df = pd.read_csv(v['file path'])\n",
    "        # check barcode\n",
    "        rename_contigs = set()\n",
    "        for i in tmp_df.index:\n",
    "            if tmp_df.at[i, 'barcode'] in rename_barcodes:\n",
    "                # rename barcode\n",
    "                tmp_df.at[i, 'barcode'] = tmp_df.at[i, 'barcode'] + '_' + k\n",
    "                # do not forget to rename contig id\n",
    "                rename_contigs.add(tmp_df.at[i, 'contig_id'])\n",
    "                tmp_df.at[i, 'contig_id'] = tmp_df.at[i, 'contig_id'] + '_' + k\n",
    "        to_write_csv.append(tmp_df)\n",
    "        \n",
    "        # read fastq\n",
    "        with open(v['seq path'], 'rt') as seq_file:\n",
    "            for record in SeqIO.parse(seq_file, 'fasta'):\n",
    "                if record.id in rename_contigs:\n",
    "                    record.id = record.id + '_' + k\n",
    "                    record.description = ''\n",
    "                    record.name = ''\n",
    "                    to_write_fastq.append(record)\n",
    "        \n",
    "    # save csv file\n",
    "    combined_df = pd.concat(to_write_csv, ignore_index=True)\n",
    "    combined_df.to_csv(os.path.join(output_folder, output_name, 'filtered_contig_annotations.csv'), index=False)\n",
    "    \n",
    "    # save fastq file\n",
    "    SeqIO.write(to_write_fastq, os.path.join(output_folder, output_name, 'filtered_contig.fasta'), 'fasta')\n",
    "        \n",
    "    return len(set(combined_df['barcode']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mergeData(noncite_folder, cite_folder, demultiplex_folder, s_type):\n",
    "    '''merge BCR or TCR data specified by s_type\n",
    "    '''\n",
    "    samples = {}\n",
    "\n",
    "    # nonCITE samples\n",
    "    for f in os.listdir(noncite_folder):\n",
    "        if os.path.isdir(os.path.join(noncite_folder, f)):\n",
    "            # the folder name has a pattern: sample name_BCR/TCR_VHT_cellranger\n",
    "            tmp = f.split('_')\n",
    "            sample_name = tmp[0]\n",
    "            sample_type = tmp[1]\n",
    "            \n",
    "            # only process BCR or TCR data\n",
    "            if not sample_type == s_type:\n",
    "                continue\n",
    "        \n",
    "            # nested dicts\n",
    "            if not sample_name in samples:\n",
    "                samples[sample_name] = {}\n",
    "        \n",
    "            samples[sample_name]['nonCITE'] = {'full sample name': sample_name+'_nonCITE',\n",
    "                                               'file path': os.path.join(noncite_folder, f, 'filtered_contig_annotations.csv'),\n",
    "                                               'seq path': os.path.join(noncite_folder, f, 'filtered_contig.fasta')}\n",
    "        \n",
    "    # Demultiplexed CITE samples\n",
    "    for f in os.listdir(demultiplex_folder):\n",
    "        if os.path.isdir(os.path.join(demultiplex_folder, f)):\n",
    "            # the folder name has a pattern: sample name_BCR/TCR_CITEx\n",
    "            sample_name, sample_type, cite_run = f.split('_')\n",
    "            \n",
    "            # only process BCR or TCR data\n",
    "            if not sample_type == s_type:\n",
    "                continue\n",
    "        \n",
    "            # nested dicts\n",
    "            if not sample_name in samples:\n",
    "                samples[sample_name] = {}\n",
    "            \n",
    "            samples[sample_name][cite_run] = {'full sample name': sample_name+'_'+cite_run,\n",
    "                                               'file path': os.path.join(demultiplex_folder, f, 'filtered_contig_annotations.csv'),\n",
    "                                               'seq path': os.path.join(cite_folder, cite_run+'_'+sample_type+'_VHT_cellranger', 'filtered_contig.fasta')}\n",
    "        \n",
    "    print('Total {} {} sampless'.format(len(samples), s_type))\n",
    "    \n",
    "    ## begin merge\n",
    "    stats = []\n",
    "\n",
    "    for sample_name, sample in samples.items():\n",
    "        # sample is still a dict\n",
    "        stats.append({'sample name': sample_name, 'data sets': len(sample)})\n",
    "        \n",
    "        # define output name\n",
    "        output_name = sample_name + '_' + s_type + '_VHT_cellranger'\n",
    "        # save data\n",
    "        if not(os.path.isdir(os.path.join(output_folder, output_name))):\n",
    "            os.mkdir(os.path.join(output_folder, output_name))\n",
    "            \n",
    "        # only 1 data set, no need to merge\n",
    "        if len(sample) == 1:\n",
    "            stats[-1]['nCellCITEOne'] = np.nan\n",
    "            stats[-1]['nCellCITETwo'] = np.nan\n",
    "            stats[-1]['Overlap2'] = 0\n",
    "            stats[-1]['Overlap3'] = 0\n",
    "            stats[-1]['OverlapCITE'] = 0\n",
    "        \n",
    "            for k, v in sample.items():\n",
    "                # save csc file\n",
    "                data = pd.read_csv(v['file path'])\n",
    "        \n",
    "            stats[-1]['nCellNonCITE'] = len(set(data['barcode']))\n",
    "            stats[-1]['nCellMerged'] = stats[-1]['nCellNonCITE']\n",
    "        \n",
    "            # save file, reanme all barcode and contig\n",
    "            saveFile(sample, set(data['barcode']), output_name) \n",
    "\n",
    "        elif len(sample) == 2:\n",
    "            # 1 CITE data and 1 nonCITE data\n",
    "            stats[-1]['Overlap3'] = 0\n",
    "        \n",
    "            for k, v in sample.items():\n",
    "                # determine CITE data and nonCITE data\n",
    "                if k.startswith('CITE'):\n",
    "                    cite_data = pd.read_csv(v['file path'])\n",
    "                else:\n",
    "                    noncite_data = pd.read_csv(v['file path'])\n",
    "        \n",
    "            stats[-1]['nCellCITEOne'] = len(set(cite_data['barcode']))\n",
    "            stats[-1]['nCellCITETwo'] = np.nan\n",
    "            stats[-1]['nCellNonCITE'] = len(set(noncite_data['barcode']))\n",
    "            stats[-1]['OverlapCITE'] = 0\n",
    "        \n",
    "            # overlap barcodes\n",
    "            overlap_barcode = set(cite_data['barcode']).intersection(set(noncite_data['barcode']))\n",
    "            stats[-1]['Overlap2'] = len(overlap_barcode)\n",
    "        \n",
    "            # save files, rename all barcodes and contigs\n",
    "            stats[-1]['nCellMerged'] = saveFile(sample, set(cite_data['barcode']).union(set(noncite_data['barcode'])), output_name)\n",
    "    \n",
    "    \n",
    "        elif len(sample) == 3:\n",
    "            # 2 CITE data and 1 nonCITE data\n",
    "            cite_data = []\n",
    "        \n",
    "            for k, v in sample.items():\n",
    "                # determine CITE data and nonCITE data\n",
    "                if k.startswith('CITE'):\n",
    "                    cite_data.append(pd.read_csv(v['file path']))\n",
    "                else:\n",
    "                    noncite_data = pd.read_csv(v['file path'])\n",
    "        \n",
    "            barcode_cite1 = set(cite_data[0]['barcode'])\n",
    "            barcode_cite2 = set(cite_data[1]['barcode'])\n",
    "            barcode_noncite = set(noncite_data['barcode'])\n",
    "        \n",
    "            stats[-1]['nCellCITEOne'] = len(barcode_cite1)\n",
    "            stats[-1]['nCellCITETwo'] = len(barcode_cite2)\n",
    "            stats[-1]['nCellNonCITE'] = len(barcode_noncite)\n",
    "        \n",
    "            # check overlap barcode\n",
    "            overlap3 = barcode_cite1.intersection(barcode_cite2).intersection(barcode_noncite)\n",
    "            stats[-1]['Overlap3'] = len(overlap3)\n",
    "            \n",
    "            # overlap between CITE\n",
    "            overlapCITE = barcode_cite1.intersection(barcode_cite2)\n",
    "            stats[-1]['OverlapCITE'] = len(overlapCITE)\n",
    "            \n",
    "            overlap2 = barcode_cite1.union(barcode_cite2).intersection(barcode_noncite)\n",
    "            stats[-1]['Overlap2'] = len(overlap2)\n",
    "        \n",
    "            # all barcode need rename\n",
    "            overlap_barcode =  barcode_cite1.union(barcode_cite2).union(barcode_noncite)\n",
    "        \n",
    "            # save files\n",
    "            stats[-1]['nCellMerged'] = saveFile(sample, overlap_barcode, output_name)\n",
    "\n",
    "    ## finish merge, show statistics\n",
    "    stats_df = pd.DataFrame(stats)\n",
    "    stats_df.set_index('sample name', inplace=True)\n",
    "    stats_df.loc['Total', :]= stats_df.sum(axis=0)\n",
    "    stats_df = stats_df[['data sets', 'nCellMerged', 'nCellNonCITE', 'nCellCITEOne', 'nCellCITETwo', 'OverlapCITE', 'Overlap2', 'Overlap3']]\n",
    "    stats_df.rename(columns={'Overlap2': 'OverlapCITEvsNonCITE', 'Overlap3': 'OverlapIn3Datasets', 'OverlapCITE': 'OverlapBetweenCITE'}, inplace=True)\n",
    "    print('Summary of merging {} data'.format(s_type))\n",
    "    display(stats_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merge BCR data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total 10 BCR sampless\n",
      "Summary of merging BCR data\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>data sets</th>\n",
       "      <th>nCellMerged</th>\n",
       "      <th>nCellNonCITE</th>\n",
       "      <th>nCellCITEOne</th>\n",
       "      <th>nCellCITETwo</th>\n",
       "      <th>OverlapBetweenCITE</th>\n",
       "      <th>OverlapCITEvsNonCITE</th>\n",
       "      <th>OverlapIn3Datasets</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NS0A</th>\n",
       "      <td>3.0</td>\n",
       "      <td>1326.0</td>\n",
       "      <td>849.0</td>\n",
       "      <td>318.0</td>\n",
       "      <td>159.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NS0B</th>\n",
       "      <td>3.0</td>\n",
       "      <td>1017.0</td>\n",
       "      <td>813.0</td>\n",
       "      <td>109.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NS1A</th>\n",
       "      <td>3.0</td>\n",
       "      <td>734.0</td>\n",
       "      <td>483.0</td>\n",
       "      <td>104.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NS1B</th>\n",
       "      <td>3.0</td>\n",
       "      <td>1464.0</td>\n",
       "      <td>1179.0</td>\n",
       "      <td>136.0</td>\n",
       "      <td>149.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TP8B</th>\n",
       "      <td>1.0</td>\n",
       "      <td>426.0</td>\n",
       "      <td>426.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TP9B</th>\n",
       "      <td>3.0</td>\n",
       "      <td>175.0</td>\n",
       "      <td>134.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TS2A</th>\n",
       "      <td>3.0</td>\n",
       "      <td>727.0</td>\n",
       "      <td>638.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TS2B</th>\n",
       "      <td>3.0</td>\n",
       "      <td>388.0</td>\n",
       "      <td>328.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TS3A</th>\n",
       "      <td>3.0</td>\n",
       "      <td>1241.0</td>\n",
       "      <td>1094.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>87.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TS3B</th>\n",
       "      <td>3.0</td>\n",
       "      <td>3256.0</td>\n",
       "      <td>2301.0</td>\n",
       "      <td>472.0</td>\n",
       "      <td>483.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <td>28.0</td>\n",
       "      <td>10754.0</td>\n",
       "      <td>8245.0</td>\n",
       "      <td>1307.0</td>\n",
       "      <td>1202.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             data sets  nCellMerged  nCellNonCITE  nCellCITEOne  nCellCITETwo  \\\n",
       "sample name                                                                     \n",
       "NS0A               3.0       1326.0         849.0         318.0         159.0   \n",
       "NS0B               3.0       1017.0         813.0         109.0          95.0   \n",
       "NS1A               3.0        734.0         483.0         104.0         147.0   \n",
       "NS1B               3.0       1464.0        1179.0         136.0         149.0   \n",
       "TP8B               1.0        426.0         426.0           NaN           NaN   \n",
       "TP9B               3.0        175.0         134.0          21.0          20.0   \n",
       "TS2A               3.0        727.0         638.0          61.0          28.0   \n",
       "TS2B               3.0        388.0         328.0          26.0          34.0   \n",
       "TS3A               3.0       1241.0        1094.0          60.0          87.0   \n",
       "TS3B               3.0       3256.0        2301.0         472.0         483.0   \n",
       "Total             28.0      10754.0        8245.0        1307.0        1202.0   \n",
       "\n",
       "             OverlapBetweenCITE  OverlapCITEvsNonCITE  OverlapIn3Datasets  \n",
       "sample name                                                                \n",
       "NS0A                        0.0                   0.0                 0.0  \n",
       "NS0B                        0.0                   0.0                 0.0  \n",
       "NS1A                        0.0                   0.0                 0.0  \n",
       "NS1B                        0.0                   0.0                 0.0  \n",
       "TP8B                        0.0                   0.0                 0.0  \n",
       "TP9B                        0.0                   0.0                 0.0  \n",
       "TS2A                        0.0                   0.0                 0.0  \n",
       "TS2B                        0.0                   0.0                 0.0  \n",
       "TS3A                        0.0                   0.0                 0.0  \n",
       "TS3B                        0.0                   2.0                 0.0  \n",
       "Total                       0.0                   2.0                 0.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mergeData(noncite_folder, cite_folder, demultiplex_folder, 'BCR')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merge TCR data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total 10 TCR sampless\n",
      "Summary of merging TCR data\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>data sets</th>\n",
       "      <th>nCellMerged</th>\n",
       "      <th>nCellNonCITE</th>\n",
       "      <th>nCellCITEOne</th>\n",
       "      <th>nCellCITETwo</th>\n",
       "      <th>OverlapBetweenCITE</th>\n",
       "      <th>OverlapCITEvsNonCITE</th>\n",
       "      <th>OverlapIn3Datasets</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NS0A</th>\n",
       "      <td>3.0</td>\n",
       "      <td>6970.0</td>\n",
       "      <td>4190.0</td>\n",
       "      <td>1839.0</td>\n",
       "      <td>941.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NS0B</th>\n",
       "      <td>3.0</td>\n",
       "      <td>1636.0</td>\n",
       "      <td>914.0</td>\n",
       "      <td>375.0</td>\n",
       "      <td>347.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NS1A</th>\n",
       "      <td>3.0</td>\n",
       "      <td>3314.0</td>\n",
       "      <td>2765.0</td>\n",
       "      <td>276.0</td>\n",
       "      <td>273.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NS1B</th>\n",
       "      <td>3.0</td>\n",
       "      <td>3673.0</td>\n",
       "      <td>2846.0</td>\n",
       "      <td>392.0</td>\n",
       "      <td>435.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TP8B</th>\n",
       "      <td>1.0</td>\n",
       "      <td>792.0</td>\n",
       "      <td>792.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TP9B</th>\n",
       "      <td>3.0</td>\n",
       "      <td>92.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TS2A</th>\n",
       "      <td>3.0</td>\n",
       "      <td>2589.0</td>\n",
       "      <td>1813.0</td>\n",
       "      <td>524.0</td>\n",
       "      <td>252.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TS2B</th>\n",
       "      <td>3.0</td>\n",
       "      <td>2299.0</td>\n",
       "      <td>1675.0</td>\n",
       "      <td>263.0</td>\n",
       "      <td>361.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TS3A</th>\n",
       "      <td>3.0</td>\n",
       "      <td>3018.0</td>\n",
       "      <td>2217.0</td>\n",
       "      <td>388.0</td>\n",
       "      <td>413.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TS3B</th>\n",
       "      <td>3.0</td>\n",
       "      <td>5780.0</td>\n",
       "      <td>2315.0</td>\n",
       "      <td>1691.0</td>\n",
       "      <td>1774.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <td>28.0</td>\n",
       "      <td>30163.0</td>\n",
       "      <td>19553.0</td>\n",
       "      <td>5784.0</td>\n",
       "      <td>4826.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             data sets  nCellMerged  nCellNonCITE  nCellCITEOne  nCellCITETwo  \\\n",
       "sample name                                                                     \n",
       "NS0A               3.0       6970.0        4190.0        1839.0         941.0   \n",
       "NS0B               3.0       1636.0         914.0         375.0         347.0   \n",
       "NS1A               3.0       3314.0        2765.0         276.0         273.0   \n",
       "NS1B               3.0       3673.0        2846.0         392.0         435.0   \n",
       "TP8B               1.0        792.0         792.0           NaN           NaN   \n",
       "TP9B               3.0         92.0          26.0          36.0          30.0   \n",
       "TS2A               3.0       2589.0        1813.0         524.0         252.0   \n",
       "TS2B               3.0       2299.0        1675.0         263.0         361.0   \n",
       "TS3A               3.0       3018.0        2217.0         388.0         413.0   \n",
       "TS3B               3.0       5780.0        2315.0        1691.0        1774.0   \n",
       "Total             28.0      30163.0       19553.0        5784.0        4826.0   \n",
       "\n",
       "             OverlapBetweenCITE  OverlapCITEvsNonCITE  OverlapIn3Datasets  \n",
       "sample name                                                                \n",
       "NS0A                        2.0                  16.0                 0.0  \n",
       "NS0B                        0.0                   1.0                 0.0  \n",
       "NS1A                        2.0                   2.0                 0.0  \n",
       "NS1B                        0.0                   6.0                 0.0  \n",
       "TP8B                        0.0                   0.0                 0.0  \n",
       "TP9B                        0.0                   0.0                 0.0  \n",
       "TS2A                        0.0                   0.0                 0.0  \n",
       "TS2B                        1.0                   0.0                 0.0  \n",
       "TS3A                        0.0                   1.0                 0.0  \n",
       "TS3B                        4.0                   8.0                 0.0  \n",
       "Total                       9.0                  34.0                 0.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mergeData(noncite_folder, cite_folder, demultiplex_folder, 'TCR')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
